File: 0767-reorganize-string/0767-reorganize-string.py
class Solution:
    def reorganizeString(self, s: str) -> str:
        counts = collections.Counter(s)
        maxFreq = max(counts.values())
        if maxFreq - 1 > sum(counts.values()) - maxFreq:
            return ""
        ans = []
        heap = []
        for k, v in counts.items():
            heap.append((-v, k))
        heapq.heapify(heap)
        c1, c2 = "", ""
        v1, v2 = 1, 1
        while heap:
            # print(heap)
            v1, c1 = heapq.heappop(heap)
            if heap:
                v2, c2 = heapq.heappop(heap)
            ans.append(c1)
            if c2:
                ans.append(c2)
            v1 += 1
            if v1 < 0:
                heapq.heappush(heap, (v1, c1))
            v2 += 1
            if v2 < 0:
                heapq.heappush(heap, (v2, c2))
            c2 = ""
        
        return "".join(ans)

File: 0767-reorganize-string/0767-reorganize-string.java
class State {
    int val;
    char c;
    State(int val, char c) {
        this.val = val;
        this.c = c;
    }
}

class Solution {
    public String reorganizeString(String s) {
        StringBuilder ans = new StringBuilder();
        Map<Character, Integer> counts = new HashMap<>();
        char[] chars = s.toCharArray();
        for(char c: chars){
            counts.put(c, counts.getOrDefault(c, 0) + 1);
        }
        int maxCount = Collections.max(counts.values());
        if(maxCount - 1 > s.length() - maxCount)
            return "";
        
        PriorityQueue<State> heap = new PriorityQueue<>((n1, n2) -> n2.val - n1.val);
        for(char key: counts.keySet()){
            heap.add(new State(counts.get(key), key));
        }
        State popped_2 = null;
        while(heap.size() > 0){
            State popped = heap.remove();
            ans.append(popped.c);
            if(heap.size() > 0)
                popped_2 = heap.remove();
            
            if(popped.val - 1 > 0){
                popped.val -= 1;
                heap.add(popped);
            }
            
            if(popped_2 != null){
                ans.append(popped_2.c);
                if(popped_2.val - 1 > 0){
                    popped_2.val -= 1;
                    heap.add(popped_2);
                }
            }
            popped_2 = null;
        }
        
        return ans.toString();
    }
}